frontend/web/components/release-pipelines/CreatePipelineStage.tsx

```diff
@@ -0,0 +1,237 @@
+import { SyntheticEvent, useEffect, useMemo, useState } from 'react'
+import { useGetEnvironmentsQuery } from 'common/services/useEnvironment'
+import InputGroup from 'components/base/forms/InputGroup'
+import Utils from 'common/utils/utils'
+import { useGetSegmentsQuery } from 'common/services/useSegment'
+import { StageActionType, StageTrigger } from 'common/types/responses'
+import Button from 'components/base/forms/Button'
+import Icon from 'components/Icon'
+import { FLAG_ACTIONS, TRIGGER_OPTIONS } from './constants'
+import { StageActionRequest, PipelineStageRequest } from 'common/types/requests'
+
+type DraftStageType = Omit<PipelineStageRequest, 'id' | 'pipeline' | 'project'>
+
+const getFlagActions = (trigger: string | undefined) => {
+  if (trigger === 'ON_ENTER') {
+    return FLAG_ACTIONS.ON_ENTER
+  }
+  return []
+}
+
+const CreatePipelineStage = ({
+  onChange,
+  onRemove,
+  projectId,
+  showRemoveButton,
+  stageData,
+}: {
+  stageData: DraftStageType
+  onChange: (stageData: DraftStageType) => void
+  projectId: string
+  showRemoveButton?: boolean
+  onRemove?: () => void
+}) => {
+  const [searchInput, setSearchInput] = useState('')
+  const [selectedAction, setSelectedAction] = useState<{
+    label: string
+    value: string
+  }>({ label: 'Select an action', value: '' })
+  const [selectedSegment, setSelectedSegment] = useState<{
+    label: string
+    value: number | null
+  }>()
+
+  const { data: environmentsData, isLoading: isEnvironmentsLoading } =
+    useGetEnvironmentsQuery(
+      {
+        projectId,
+      },
+      { skip: !projectId },
+    )
+
+  const { data: segments, isLoading: isSegmentsLoading } = useGetSegmentsQuery(
+    {
+      include_feature_specific: true,
+      page_size: 1000,
+      projectId,
+    },
+    { skip: !projectId },
+  )
+
+  const selectedTrigger = useMemo(() => {
+    return TRIGGER_OPTIONS.find(
+      (trigger) => trigger.value === stageData.trigger.trigger_type,
+    )
+  }, [stageData.trigger.trigger_type])
+
+  const segmentOptions = useMemo(() => {
+    return segments?.results?.map((segment) => ({
+      label: segment.name,
+      value: segment.id,
+    }))
+  }, [segments])
+
+  const environmentOptions = useMemo(() => {
+    return environmentsData?.results?.map((environment) => ({
+      label: environment.name,
+      value: environment.id,
+    }))
+  }, [environmentsData])
+
+  const handleOnChange = (
+    fieldName: keyof DraftStageType,
+    value: string | number | StageTrigger | StageActionRequest[],
+  ) => {
+    onChange({ ...stageData, [fieldName]: value })
+  }
+
+  useEffect(() => {
+    if (environmentOptions?.length && stageData.environment === -1) {
+      handleOnChange('environment', environmentOptions?.[0]?.value)
+    }
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [environmentOptions, stageData])
+
+  useEffect(() => {
```

Comment:
As those useEffects are side effects triggered by changes in single values.
Couldn't we simplify and directly have a handler for each of the input ? `onSelectSegment`, `onSelectAction` ?
I believe it would have the same results without the useEffects risks ?